File: 127-wordLadder/solution.java
class Solution {
    public int ladderLength(String beginWord, String endWord, List<String> wordList) {
        /*
        // 92% least-nodes-first bi-directional BFS
        Set<String> dict = new HashSet<>(wordList);
        Set<String> startQ = new HashSet<>();
        Set<String> endQ = new HashSet<>();
        Set<String> visited = new HashSet<>();
        
        if(!dict.contains(endWord)){
            return 0;
        }
        
        startQ.add(beginWord);
        endQ.add(endWord);
        
        for(int leng = 2;!startQ.isEmpty();leng++){
            Set<String> neighbours = new HashSet<>();
            
            //go through all the nodes in the queue
            for(String node: startQ){                
                //for each node, visit their neighbouring nodes
                for(int i=0;i<node.length();i++){
                    StringBuilder transformWord = new StringBuilder(node);
                    char exclude = transformWord.charAt(i);
                    for(char c='a';c<='z';c++){
                        if(c == exclude){
                            continue;
                        }
                        transformWord.setCharAt(i,c);
                        String tWord = transformWord.toString();

                        if(endQ.contains(tWord)){
                            return leng;
                        }

                        if(dict.contains(tWord) && visited.add(tWord)){
                            neighbours.add(tWord);
                        }
                    }
                }
            }
            
            //by here, we have processed all of previous queue
            //Choose the shortest between the startQ and endQ
            //in hopes to alternate between them to meet somewhere
            //at the middle. This optimizes the code, because we are processing 
            //smallest queue first, so # of nodes dont blow up as fast.
            //basically balancing between the two queues
            startQ = (neighbours.size() > endQ.size()) ? endQ : neighbours;
            endQ = (startQ == neighbours) ? endQ : neighbours;
        }
        
        return 0;
        */

        // 45% BFS
        List<String> taskQ = new ArrayList();
        taskQ.add(beginWord);
        Map<String, Integer> dict = new HashMap<String, Integer>();
        for(String s: wordList){
            dict.put(s, 0);
        }
        int path =1;

        while(taskQ.size()>0){

            int size=taskQ.size();
            List<String> neighbours = new ArrayList();
            while(size>0){

                size--;
                String word = taskQ.remove(0);
                // System.out.println("1 inin "+word);

                for(int i=0; i<word.length(); i++){
                    StringBuilder transformWord = new StringBuilder(word);

                    for(char alphabet = 'a'; alphabet <='z'; alphabet++ ){
                        transformWord.setCharAt(i,alphabet);
                        String tWord = transformWord.toString();
                        
                        
                        if (dict.get(tWord)==null || dict.get(tWord)==1){
                            continue;
                        }
                        if (tWord.equals(endWord)){
                            return path+1;
                        }
                        neighbours.add(tWord);
                        dict.put(tWord, 1);
                    }
                }
            }
            for(String s: neighbours) {
                // System.out.print(s+", ");
                taskQ.add(s);
            }
            // System.out.println(" ");
            path+=1;
        }
        return 0;
    }
}

File: 127-wordLadder/solution.py
import string

class Solution:
    def ladderLength(self, beginWord, endWord, wordList):
        """
        :type beginWord: str
        :type endWord: str
        :type wordList: List[str]
        :rtype: int
        """
        # 30%, acceptable solution
       
        taskQ = [beginWord]
        word_list = {i: 0 for i in wordList}
        path = 1  
        
        while len(taskQ)>0:
            size = len(taskQ)
            
            candidates=[]
            
            while size>0:
                size-=1
                word = taskQ.pop(0)
                
                for i in range(len(word)):
                    for letter in 'abcdefghijklmnopqrstuvwxyz':
                        w = word[:i]+letter+word[i+1:]
                       
                        if word_list.get(w, -1)<0:
                            continue
                        if word_list[w]:
                            continue
                        
                        if w == endWord:
                            return path+1
                        candidates.append(w)
                        word_list[w]=1
                
            candidates = list(set(candidates))
            taskQ.extend(candidates)
                
            path+=1
        return 0
    '''
    # 80% using set (bigO(1) when checking exist) & bi-direction BFS
        word_dict =set(wordList)
        fQ = [beginWord]
        bQ = [endWord]
        fCount, bCount = 0, 1  
        fInd, bInd = 0, 0
        fRecord=set()
        bRecord=set()

        if not(endWord in word_dict): return 0
        while True:
            # first part ~~ 25%
            if len(fQ)<=len(bQ):
                candidates=[]

                for i in range(fInd, len(fQ)):
                    word = fQ[i]
                    
                    for i in range(len(word)):
                        for letter in 'abcdefghijklmnopqrstuvwxyz':
                            w = word[:i]+letter+word[i+1:]
                            if not(w in word_dict): continue
                            if w in fRecord: continue
                            
                            candidates.append(w)
                            fRecord.add(w)

                candidates = list(set(candidates))
                if len(candidates)==0: break
                fInd=len(fQ)
                fQ.extend(candidates)
                fCount+=1
                if set(fQ) & set(bQ):
                    return fCount+bCount
            else:
                candidates=[]

                for i in range(bInd, len(bQ)):
                    word = bQ[i]
                    
                    for i in range(len(word)):
                        for letter in 'abcdefghijklmnopqrstuvwxyz':
                            w = word[:i]+letter+word[i+1:]
                            if not(w in word_dict): continue
                            if w in bRecord: continue
                            
                            candidates.append(w)
                            bRecord.add(w)
    
                candidates = list(set(candidates))
                if len(candidates)==0: break
                bInd=len(bQ)
                bQ.extend(candidates)
                bCount+=1
                if set(fQ) & set(bQ):
                    return fCount+bCount
            
        return 0
    '''

    '''
    # 90%, bi-direction BFS with less nodes
    word_dict = set(wordList)
        startQ = set()
        endQ = set()
        visited = set()
        
        if endWord not in word_dict:
            return 0
        
        startQ.add(beginWord)
        endQ.add(endWord)
        leng = 2
        while len(startQ) != 0:
            neighbours = set()
            for node in startQ:
                for i in range(0,len(node)):
                    char_arr = list(node)
                    exclude = ord(node[i]);
                    for j in range(ord('a'),ord('z')+1):
                        if j == exclude:
                            continue
                        
                        char_arr[i] = chr(j)
                        transformWord = ''.join(char_arr)
                        
                        if transformWord in endQ:
                            return leng
                        
                        if transformWord in word_dict and transformWord not in visited:
                            visited.add(transformWord)
                            neighbours.add(transformWord)
            
            startQ = endQ if len(endQ) < len(neighbours) else neighbours
            endQ = endQ if startQ == neighbours else neighbours
            leng += 1
        
        return 0
        '''   